src/app/clusters/webrtc-transport-requestor-server/webrtc-transport-requestor-cluster.cpp

```diff
@@ -0,0 +1,417 @@
+
+        if (FindSession(candidateId) == nullptr)
+        {
+            break;
+        }
+    }
+
+    return candidateId;
+}
+
+bool WebRTCTransportRequestorServer::IsPeerNodeSessionValid(uint16_t sessionId, const CommandHandler & commandHandler)
+{
+    NodeId peerNodeId           = GetNodeIdFromCtx(commandHandler);
+    FabricIndex peerFabricIndex = commandHandler.GetAccessingFabricIndex();
+
+    // Check if the session ID is in the existing sessions list
+    WebRTCSessionStruct * existingSession = FindSession(sessionId);
+
+    if (!existingSession)
+    {
+        return false;
+    }
+
+    // Also check that the existing session belongs to the same PeerNodeID / Fabric
+    // If it doesn't match, return false
+    if (peerNodeId != existingSession->peerNodeID || peerFabricIndex != existingSession->GetFabricIndex())
+    {
+        return false;
+    }
+
+    return true;
+}
+
+DataModel::ActionReturnStatus WebRTCTransportRequestorServer::HandleOffer(const Commands::Offer::DecodableType & req,
+                                                                          const CommandHandler & commandHandler)
+{
+    // Extract command fields from the request
+    uint16_t sessionId = req.webRTCSessionID;
+
+    // Check if the session, NodeID are valid
+    if (!IsPeerNodeSessionValid(sessionId, commandHandler))
+    {
+        return Protocols::InteractionModel::Status::NotFound;
+    }
+
+    // Create arguments for WebRTCTransportRequestorDelegate.
+    WebRTCTransportRequestorDelegate::OfferArgs args;
+    args.sdp = std::string(req.sdp.data(), req.sdp.size());
+
+    // Convert ICE servers list from DecodableList to vector.
+    if (req.ICEServers.HasValue())
+    {
+        auto iterator = req.ICEServers.Value().begin();
+        while (iterator.Next())
+        {
+            args.iceServers.Value().push_back(iterator.GetValue());
+        }
+
+        // Check the list validity
+        CHIP_ERROR err = iterator.GetStatus();
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(Zcl, "HandleOffer: ICECandidates list error: %" CHIP_ERROR_FORMAT, err.Format());
+            return Protocols::InteractionModel::Status::InvalidCommand;
+        }
+    }
+
+    // Convert ICETransportPolicy from CharSpan to std::string.
+    if (req.ICETransportPolicy.HasValue())
+    {
+        args.iceTransportPolicy.SetValue(std::string(req.ICETransportPolicy.Value().data(), req.ICETransportPolicy.Value().size()));
+    }
+
+    // Delegate processing: handle the SDP offer, gather ICE candidates, SDP answer, etc.
+    CHIP_ERROR delegateResult = mDelegate.HandleOffer(sessionId, args);
+    if (delegateResult == CHIP_NO_ERROR)
+    {
+        return Protocols::InteractionModel::Status::Success;
+    }
+    else if (delegateResult == CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE)
+    {
+        return Protocols::InteractionModel::Status::UnsupportedCommand;
+    }
+    else
+    {
+        return Protocols::InteractionModel::Status::Failure;
```

Comment:
ActionReturnStatus can cast from CHIP_ERROR (and the general fallback is Failure). I would return the chip_error here instead of forcibly converting to failure.